File: Difficulty: Medium/Smaller on Left/smaller-on-left.js
//{ Driver Code Starts
//Initial Template for javascript

"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;

process.stdin.on("data", (inputStdin) => {
  inputString += inputStdin;
});

process.stdin.on("end", (_) => {
  inputString = inputString
    .trim()
    .split("\n")
    .map((string) => {
      return string.trim();
    });

  main();
});

function readLine() {
  return inputString[currentLine++];
}
function printArray(arr) {
  let s = "";
  for (let i of arr) {
    s = s + i + " ";
  }
  console.log(s);
}

function main() {
  let t = parseInt(readLine());
  let i = 0;
  for (; i < t; i++) {
    let n = parseInt(readLine());
    let arr = readLine()
      .trim()
      .split(" ")
      .map((x) => parseInt(x));
    let obj = new Solution();
    let res = obj.leftSmaller(arr, n);
    printArray(res);
  
console.log("~");
}
}
// } Driver Code Ends


//User function Template for javascript

/**
 * @param {number[]} arr
 * @param {number} n
 * @return {number[]}
 */
class Solution {
  leftSmaller(arr, n) {
    // code here
    let result = [];  // Array to store the nearest smaller element for each index
    let stack = [];  // Stack to keep track of elements in the array while processing

    // Traverse the array
    for (let i = 0; i < n; i++) {
        // Pop elements from stack until we find one smaller than the current element
        while (stack.length && stack[stack.length - 1] >= arr[i]) {
            stack.pop();
        }

        // If the stack is empty, no smaller element found, so assign -1
        if (!stack.length) result[i] = -1;
        else result[i] = stack[stack.length - 1];  // The top of the stack is the nearest smaller element

        stack.push(arr[i]);  // Push the current element onto the stack for future comparison
    }

    return result;  // Return the result array
  }
}

File: Difficulty: Medium/Smaller on Left/smaller-on-left.java
//{ Driver Code Starts
// Initial Template for Java

// Initial Template for Java

// Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;

class Array {

    // Driver code
    public static void main(String[] args) throws IOException {
        // Taking input using buffered reader
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int testcases = Integer.parseInt(br.readLine());

        // looping through all testcases
        while (testcases-- > 0) {
            int sizeOfArray = Integer.parseInt(br.readLine());
            int arr[] = new int[sizeOfArray];

            String line = br.readLine();
            String[] elements = line.trim().split("\\s+");
            for (int i = 0; i < sizeOfArray; i++) {
                arr[i] = Integer.parseInt(elements[i]);
            }

            Complete obj = new Complete();
            arr = obj.Smallestonleft(arr, sizeOfArray);
            for (int i = 0; i < sizeOfArray; i++) {
                System.out.print(arr[i] + " ");
            }
            System.out.println();
            // System.out.println(re);
        }
    }
}

// } Driver Code Ends


// User function Template for Java
class Complete {

    // Function for finding the maximum element smaller than the current element on its left
    public static int[] Smallestonleft(int arr[], int n) {
        // Create a TreeSet to store the elements in sorted order
        TreeSet<Integer> set = new TreeSet<>();
        
        // Array to store the result
        int res[] = new int[n];
        
        // The first element has no smaller element on its left, so assign -1
        res[0] = -1;
        
        // Add the first element to the TreeSet
        set.add(arr[0]);
        
        // Traverse through the rest of the elements in the array
        for (int i = 1; i < n; i++) {
            // Find the largest element in the set that is smaller than arr[i]
            Integer s = set.lower(arr[i]);
            
            // If such an element exists, assign it; otherwise, assign -1
            res[i] = (s != null) ? s : -1;
            
            // Add the current element to the set for future comparisons
            set.add(arr[i]);
        }
        
        return res;
    }
}
